pkg/ingester/retention.go

```diff
@@ -39,183 +65,322 @@ func defaultRetentionPolicy() retentionPolicy {
-
-	stopCh chan struct{}
-	wg     sync.WaitGroup
-}
-
-type tenantBlock struct {
-	ulid     ulid.ULID
-	tenantID string
-	path     string
-}
-
-type fileSystem interface {
-	fs.ReadDirFS
-	RemoveAll(string) error
+type retentionPolicy struct {
+	MinFreeDisk                uint64
+	MinDiskAvailablePercentage float64
+	EnforcementInterval        time.Duration
 }
 
-type realFileSystem struct{}
-
-func (*realFileSystem) Open(name string) (fs.File, error)          { return os.Open(name) }
-func (*realFileSystem) ReadDir(name string) ([]fs.DirEntry, error) { return os.ReadDir(name) }
-func (*realFileSystem) RemoveAll(path string) error                { return os.RemoveAll(path) }
+// diskCleaner monitors disk usage and cleans unused data.
+type diskCleaner struct {
+	services.Service
 
-// blockEvicter unloads blocks from tenant instance.
-type blockEvicter interface {
-	// evictBlock evicts the block by its ID from the memory and
-	// invokes fn callback, regardless of if the tenant is found.
-	// The call is thread-safe: tenant can't be added or removed
-	// during the execution.
-	evictBlock(tenant string, b ulid.ULID, fn func() error) error
-}
+	logger        log.Logger
+	config        phlaredb.Config
+	policy        retentionPolicy
+	blockManager  fsBlockManager
+	volumeChecker diskutil.VolumeChecker
 
-func newRetentionPolicyEnforcer(logger log.Logger, blockEvicter blockEvicter, retentionPolicy retentionPolicy, dbConfig phlaredb.Config) *retentionPolicyEnforcer {
-	e := retentionPolicyEnforcer{
-		logger:          logger,
-		blockEvicter:    blockEvicter,
-		retentionPolicy: retentionPolicy,
-		dbConfig:        dbConfig,
-		stopCh:          make(chan struct{}),
-		fileSystem:      new(realFileSystem),
-		volumeChecker:   diskutil.NewVolumeChecker(retentionPolicy.MinFreeDisk*1024*1024*1024, retentionPolicy.MinDiskAvailablePercentage),
-	}
-	e.Service = services.NewBasicService(nil, e.running, e.stopping)
-	return &e
+	stop chan struct{}
+	wg   sync.WaitGroup
 }
 
-func (e *retentionPolicyEnforcer) running(ctx context.Context) error {
-	e.wg.Add(1)
-	retentionPolicyEnforcerTicker := time.NewTicker(e.retentionPolicy.EnforcementInterval)
+func (dc *diskCleaner) running(ctx context.Context) error {
+	dc.wg.Add(1)
+	ticker := time.NewTicker(dc.policy.EnforcementInterval)
 	defer func() {
-		retentionPolicyEnforcerTicker.Stop()
-		e.wg.Done()
+		ticker.Stop()
+		dc.wg.Done()
 	}()
+
+	var deleted int
+	var bytesDeleted int
+	var hasHighDiskUtilization bool
 	for {
-		// Enforce retention policy immediately at start.
-		level.Debug(e.logger).Log("msg", "enforcing retention policy")
-		if err := e.cleanupBlocksWhenHighDiskUtilization(ctx); err != nil {
-			level.Error(e.logger).Log("msg", "failed to enforce retention policy", "err", err)
+		deleted = dc.DeleteUploadedBlocks(ctx)
+		level.Debug(dc.logger).Log("msg", "cleaned uploaded blocks", "count", deleted)
+
+		deleted, bytesDeleted, hasHighDiskUtilization = dc.EnforceHighDiskUtilization(ctx)
```

Comment:
If I read it right, this can remove blocks that were uploaded but not yet discovered by store-gateway. This means that deleted blocks won't be visible to queriers for some period of time (util store-gateway discovers them). This is fine, if we have to do so due to lack of free space, and the removal is vital. But it seems we remove _all_ uploaded blocks unconditionally.

As of now, when we query blocks, we rely on the `-querier.query-store-after` parameter (but I hope we will get rid of them at some point):

```yaml
# The time after which a metric should be queried from storage and not just
# ingesters. 0 means all queries are sent to store. If this option is enabled,
# the time range of the query sent to the store-gateway will be manipulated to
# ensure the query end is not more recent than 'now - query-store-after'.
# CLI flag: -querier.query-store-after
[query_store_after: <duration> | default = 4h]
```

Theoretically, we could rely here on this parameter as well, or check that at least 2 sync intervals passed since the block has been created, like:
```
now() - block.timestamp > 2 * sg-sync-interval
```
This isn't a reliable approach, and I'd rather try to avoid such speculations. But there is currently no simple (or efficient at least) way to confidently say that store-gateways are aware of the block and it will be read on query.

Another thing that we should keep in mind is that some of the users won't ever use the object store.

This makes me thinking that we, probably, should introduce a new parameter, that would explicitly control for how long uploaded blocks are kept locally (sort of time-based retention policy for local blocks).

Reply:
Agreed on the configuration parameter. It's unfortunate we have to add such a configuration, but this should allow folks to change how long to keep shipped blocks around. They can make their own trade offs (i.e. if the parameter is set too low, there may be query gaps before store gateways sync shipped blocks, too high and too much disk may be used).

I made this change in https://github.com/grafana/pyroscope/pull/2816/commits/2a763aff9489bd6ea2294d410d977cc45c06191d.

Reply:
Basically as long as a block is older than `now() - expiry`, it will be deleted. By default, the expiry is twice the length of the bucket store sync interval (30 min).

Reply:
That said, the `EnforceHighDiskUtilization` function will not respect this. It will try first delete the blocks marked as shipped (oldest to newest), then move to the unshipped blocks (oldest to newest).

Reply:
Yeah, that looks good!